src/test/generate-sym-test-inv.sh

```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+set -eux
+
+TARGET_DIR="${1:?}"
+
+cat <<EOF
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+const struct {
+        const char *name;
+        const void *symbol;
+} symbols_from_source[] = {
+EOF
+
+# functions
+grep -r '^_public_' "$TARGET_DIR" | sed -ne '/[ *][0-9a-zA-Z_]*(/ { s/^.*[ *]\([0-9a-zA-Z_]*\)(.*$/        { "\1", \1 },/; p }'
+# variables
+grep -r '^_public_' "$TARGET_DIR" | sed -ne '/[ *][0-9a-zA-Z_]* =/ { s/^.*[ *]\([0-9a-zA-Z_]*\) =.*$/        { "\1", \&\1 },/; p }'
+# ref/unref functions defined through macro
+grep -r '^DEFINE_PUBLIC_TRIVIAL_REF_FUNC' "$TARGET_DIR" | sed -e 's/.*DEFINE_PUBLIC_TRIVIAL_REF_FUNC([^,]*, *\([0-9a-zA-Z_]*\) *).*$/        { "\1_ref", \1_ref },/'
+grep -r '^DEFINE_PUBLIC_TRIVIAL_UNREF_FUNC' "$TARGET_DIR" | sed -e 's/.*DEFINE_PUBLIC_TRIVIAL_UNREF_FUNC([^,]*, *\([0-9a-zA-Z_]*\),.*$/        { "\1_unref", \1_unref },/'
+grep -r '^DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC' "$TARGET_DIR" | sed -e 's/.*DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC([^,]*, *\([0-9a-zA-Z_]*\),.*$/        { "\1_ref", \1_ref },\n        { "\1_unref", \1_unref },/'
+
```

Comment:
If you want, you can "prettify" this a bit using awk:

```bash
#!/usr/bin/env bash
# SPDX-License-Identifier: LGPL-2.1-or-later

set -eux

TARGET_DIR="${1:?}"

cat <<EOF
/* SPDX-License-Identifier: LGPL-2.1-or-later */
#pragma once
const struct {
        const char *name;
        const void *symbol;
} symbols_from_source[] = {
EOF

while read -r file; do
    awk '
    # Functions
    match($0, /^_public_\s+([^\S]+\s+)+\*?(\w+)\s*\(/, m) {
        printf "        { \"%s\", %s },\n", m[2], m[2]
    }
    # Variables
    match($0, /^_public_\s+([^\S]+\s+)+\*?(\w+)\s*=/, m) {
        printf "        { \"%s\", &%s },\n", m[2], m[2]
    }
    # Functions defined through a macro
    match($0, /^DEFINE_PUBLIC_TRIVIAL_REF_FUNC\([^,]+,\s*(\w+)\s*,/, m) {
        printf "        { \"%s_ref\", %s_ref },\n", m[1], m[1]
    }
    match($0, /^DEFINE_PUBLIC_TRIVIAL_UNREF_FUNC\([^,]+,\s*(\w+)\s*,/, m) {
        printf "        { \"%s_unref\", %s_unref },\n", m[1], m[1]
    }
    match($0, /^DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC\([^,]+,\s*(\w+)\s*,/, m) {
        printf "        { \"%s_ref\", %s_ref },\n", m[1], m[1]
        printf "        { \"%s_unref\", %s_unref },\n", m[1], m[1]
    }
    ' "$file"

done < <(find "$TARGET_DIR" -type f)

cat <<EOF
        {}
};
EOF
```

Definitely not required and it has definitely nothing to do with my affection to awk :)

Reply:
Thanks.

Reply:
Hm... it does not work with mawk, and gawk is not installed by default in debian and ubuntu.

Reply:
Ah, beans, I always forget about this. Well, in that case I'd just write in python, e.g.:

```python
#!/usr/bin/env python3

import os
import sys
import re

def process_file(file):
    for line in file:
        # Functions
        m = re.search(r"^_public_\s+(\S+\s+)+\**(\w+)\s*\(", line)
        if m:
            print("        {{ \"{0}\", {0} }},".format(m[2]))
        # Variables
        m = re.search(r"^_public_\s+(\S+\s+)+\**(\w+)\s*=", line)
        if m:
            print("        {{ \"{0}\", &{0} }},".format(m[2]))
        # Functions defined through a macro
        m = re.search(r"^DEFINE_PUBLIC_TRIVIAL_REF_FUNC\([^,]+,\s*(\w+)\s*,", line)
        if m:
            print("        {{ \"{0}_ref\", {0}_ref }},".format(m[1]))
        m = re.search(r"^DEFINE_PUBLIC_TRIVIAL_UNREF_FUNC\([^,]+,\s*(\w+)\s*,", line)
        if m:
            print("        {{ \"{0}_unref\", {0}_unref }},".format(m[1]))
        m = re.search(r"^DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC\([^,]+,\s*(\w+)\s*,", line)
        if m:
            print("        {{ \"{0}_ref\", {0}_ref }},".format(m[1]))
            print("        {{ \"{0}_unref\", {0}_unref }},".format(m[1]))

def main():
    print("""
/* SPDX-License-Identifier: LGPL-2.1-or-later */
#pragma once
const struct {
        const char *name;
        const void *symbol;
} symbols_from_source[] = {
    """)

    for dirpath, _, filenames in os.walk(sys.argv[1]):
        for filename in filenames:
            with open(os.path.join(dirpath, filename), "r") as f:
                process_file(f)

    print("""
        {}
};
    """)

if __name__ == "__main__":
    main()
```

this should work fine with python3.5+ and we already have a dep on python3.5.